activerecord/lib/active_record/associations/builder/belongs_to.rb

```diff
@@ -114,9 +114,13 @@ def self.add_touch_callbacks(model, reflection)
         BelongsTo.touch_record(record, record.send(changes_method), foreign_key, n, touch, belongs_to_touch_method)
       }}
 
-      model.after_save    callback.(:saved_changes), if: :saved_changes?
-      model.after_touch   callback.(:changes_to_save)
-      model.after_destroy callback.(:changes_to_save)
+      unless reflection.counter_cache_column
+        model.after_create callback.(:saved_changes), if: :saved_changes?
+        model.after_destroy callback.(:changes_to_save)
```

Comment:
I have a change in behavior which I believe is associated with removing the after_destroy callback for items with counter caches.

In 5.1.4, models chained together with belongs_to and touch: true would changed their updated_at timestamps regardless of the presence of a counter cache. Now, a model with a counter_cache won't propagate the touch past the next model.

Example:

```
class Books
    belongs_to :author, touch: true, counter_cache: true
end

class Author
  belongs_to :publisher, touch: true
end
```

In 5.1.4, when you destroy a Book, the Author and the Publisher both get new updated_at values.

In 5.1.5, only the Author is updated.

Reply:
I just ran into this as well when migrating to 5.1.6.1. How was this not considered a breaking change?

Reply:
Could you try `5-1-stable` branch (because `5.1.7` hasn't been released yet). I suppose it should be fixed there by https://github.com/rails/rails/commit/81dd13dfcea75ea7e65e52adfe5caed8868f5aae?

Reply:
I upgraded to 5.1.7 but still experiencing this.